Add ActivationPolicy= [Link] parameter

ddstreet opened the pull request:
This parameter controls when networkd will adjust the interface IFF_UP flag.  It allows always keeping the interface either up or down, or just initially taking the interface up or down at networkd startup, or not modifying IFF_UP at all, to allow fully manual control.

This is related to #15060

poettering commented on man/systemd.network.xml:
```diff
@@ -244,6 +244,38 @@
+      <varlistentry>
+        <term><varname>AdministrativeState=</varname></term>
+        <listitem>
+          <para>Takes a string indicating how networkd should manage the link administrative state.
```
`networkd` → `<command>systemd-networkd</command>`

ddstreet replied:
ack, changed

poettering commented on man/systemd.network.xml:
```diff
@@ -244,6 +244,38 @@
+          When <literal>manual</literal>, networkd will not change the administrative state at all;
+          the system administrator must bring the interface administratively up or down, as desired.
+          When <literal>start-on</literal> or <literal>start-off</literal>, networkd will set the
+          administrative state on or off, respectively, when networkd starts, but it will not change
```
ditto (and everywhere else too)

(Yeah, I know we are not very good at this right now, but let's try to get this right at least for new documentation)

poettering commented on man/systemd.network.xml:
```diff
@@ -244,6 +244,38 @@
+
+          <para>Note if <literal>always-on</literal> or <literal>always-off</literal> are used,
+          the administrative state can still be manually changed, but networkd will always
+          immediately set the administrative state back to its previous state.  This state bounce
```
previous state → state configured with this option

poettering replied:
"state bounce"? Is that established terminology? I at least (as a non-native English speaker) don#t know the term... Maybe call this "brief state change"

ddstreet replied:
i reworded this to try to make it clearer, and dropped the 'bounce', let me know if it's better

poettering commented on man/systemd.network.xml:
```diff
@@ -244,6 +244,38 @@
+          <para>This parameter conflicts with <varname>BindCarrier=</varname>, since that
+          parameter binds the interface's administrative state to the carrier state of other
+          interface(s). If both are set, a warning will be logged and
+          <varname>AdministrativeState=</varname> will be ignored.</para>
```
maybe introduce an explicit state for this called "bound" or so? And BindCarrier= would imply that the state is set to "bound"?

This would be useful when querying the state, as then BindCarrier= would be easily discoverable just by looking at the state

ddstreet replied:
great idea! i updated it to include a 'bound' state that is required/forced with BindCarrier=

ddstreet commented:
> I am a bit concerned about the naming conflict with the D-Bus property "AdministrativeState".
> 
> I think this new added feature probably should be called "State" anyway, but "Policy" or so. Aftr all state is what is influenced by this setting, it declares when to change state, but the state itself is independent.
> 
> Maybe call this ActivationPolicy=?

policy would make more sense for the naming, though 'activation' doesn't really feel right, though 'AdministrativePolicy' doesn't really sound much better either.

I'll address your other comments and update the name to ActivationPolicy, unless I can think of anything else to suggest :)

ddstreet commented:
> > Maybe call this ActivationPolicy=?

yep, i changed it to ActivationPolicy=, it sounds ok to me now :)

the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.

TriMoon commented:
@ddstreet 
> the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.

IMHO, a better and more to-the-point naming would be "**initial-up**" and "**initial-down**" as that is what they are supposed to do right?
Those names also make more sense with the "Policy" rename....

ddstreet commented:
> @ddstreet
> 
> > the possible values are start-up, always-up, manual, always-down, start-down, and bound; however i wonder if the 'start-up' and 'start-down' should just be 'up' and 'down'? the 'start-' seems redundant, maybe. i can change it and re-push if you think 'up' and 'down' sound better than 'start-up' and 'start-down'.
> 
> IMHO, a better and more to-the-point naming would be "**initial-up**" and "**initial-down**" as that is what they are supposed to do right?
> Those names also make more sense with the "Policy" rename....

initial-up/initial-down do sound better and more accurate than start-up/start-down, but just plain 'up' and 'down' still seem better to me (and the remaining possible values would still be 'always-up', 'always-down', 'manual' and 'bound'), though i have no strong feelings; i'll wait for @poettering opinion before changing the PR to any new names.

poettering commented:
BTW, You can add "Fixes: #3031" to your commits

keszybz commented on man/systemd.network.xml:
```diff
@@ -228,6 +228,39 @@
+          <literal>start-down</literal>, <command>systemd-networkd</command> will set the link
+          admin state up or down, respectively, when <command>systemd-networkd</command> starts
+          (or, when the interface is first detected, if that occurs after <command>systemd-networkd</command>
+          starts), but it will not change the flag after the first time. When <literal>always-up</literal>
```
Hmm, networkd now supports configuration reloads, so it may bring interfaces up not only at start. So instead of referring to networkd startup, say "... will set the administrative state of the link to up or down after configuring the interface, but will not change the state afterwards".

Also, with "start-up", if the interface is manually set to down, will `networkctl reconfigure` bring it back up? I think it should.

ddstreet replied:
I updated the doc, to include clarifying it 'up' and 'down' will apply only during (re)configuration; let me know if you think the new wording should be changed.

> Also, with "start-up", if the interface is manually set to down, will networkctl reconfigure bring it back up? I think it should.

I agree; I moved the flag out of the link and into the link->network, so up/down policy should be applied each reconfiguration.

ddstreet replied:
moving the ->activated flag into link->network was a mistake on my part (since ->network can be shared between links and shouldn't hold link state, i'm pretty sure this was causing the test failures), but i just moved it back to link->activated and it should still be reset on networkctl reload, since that will call link_reconfigure() with force=True.

keszybz commented on man/systemd.network.xml:
```diff
@@ -228,6 +228,39 @@
+          connection. However, to avoid reconfiguration failures, when using <literal>always-up</literal>,
+          <varname>IgnoreCarrierLoss=</varname> is forced to true.</para>
+
+          <para>The default is <literal>start-up</literal>.</para>
```
I think this para should be moved before the previous one.

ddstreet replied:
ack, i moved it to "When <literal>up</literal> (the default)"

keszybz reviewed (commented):
IIUC, this changes the behaviour slightly: the new default is "start-up", and if I'm reading the diff correctly, the code path that was taken previously corresponds to the "always-up" option. Can you please clarify (in the commit message) if there's any behaviour change.

ddstreet commented:
Sorry for the delay in responding. tl;dr: i'll update the commit msg to clarify if this changes previous/default behavior.

> IIUC, this changes the behaviour slightly: the new default is "start-up", and if I'm reading the diff correctly, the code path that was taken previously corresponds to the "always-up" option. Can you please clarify (in the commit message) if there's any behaviour change.

So, this is one reason I originally felt that 'ActivationPolicy' was not as clear as 'AdministrativeState'.  It's true that the previous policy had systemd-networkd setting the IFF_UP flag bit to 1 (if not already 1) every time link_joined() is called, however, once IFF_UP is set to 1, it should *never* change back to 0, unless some administrator explicitly sets it to 0. Losing carrier won't set IFF_UP to 0, nor will any address or routing changes. The IFF_UP flag is essentially like the physical key to a car; the car doesn't necessarily start and/or run once you turn the key 'on', but it *definitely* will not start/run without the key 'on'.

As the previous behavior was to set IFF_UP to 1 anytime link_joined() is called, it doesn't exactly match the 'always-up' setting; previously, if an admin set IFF_UP to 0, systemd-networkd would notice the flag change, but do nothing; then since losing IFF_UP causes carrier to be lost, systemd-networkd would also notice carrier loss and (unless IgnoreCarrierLoss=true) would unconfigure the interface. However, since carrier would *never* be regained until some adimn manually set IFF_UP back to 1, systemd-networkd would never do anything else for the interface, and once the admin sets IFF_UP back to 1, systemd-networkd doesn't need to set IFF_UP to 1 itself (since the admin just did it manually).

However; with the new addition of 'networkctl up/down', i should take a look at this again.

ddstreet commented on src/network/networkd-link.c:
```diff
@@ -4089,6 +4110,16 @@ int link_carrier_reset(Link *link) {
+                return 0;
+
+        if (link->network->activation_policy == ACTIVATION_POLICY_ALWAYS_DOWN) {
+                log_link_info(link, "ActivationPolicy is always off, forcing link down");
```
ack, will change this, thanks.

yuwata commented:
Sorry, I made a large change on networkd. Could you rebase this?

ddstreet commented:
I split the first patch from this out, as it is fairly simple and can be considered separately, in #17392 (since it's still needed in this PR, it is still included here too)

I also dropped the patch that just added a comment about the 'AdministrativeState' dbus event, as that's not really needed/important now that this is named 'ActivationPolicy' (though I do still think the dbus event is named confusingly, at least to people familiar with the older networking 'administrative state' terminology)

I rebased and pushed just so it will be up to date, but I want to review the changes again, and I might need to adjust and push again

poettering commented on src/network/networkd-link.h:
```diff
@@ -128,6 +128,7 @@ typedef struct Link {
         bool setting_genmode:1;
         bool ipv6_mtu_set:1;
         bool bridge_mdb_configured:1;
+        bool started:1;
```
i'd appreciate a quick comment added here that explains when this is set precisely

ddstreet replied:
yep sorry, will update again today to address this and the other previous comments

poettering reviewed (changes requested):
I think @keszybz' comments werent' addressed yet.

ddstreet commented:
Ok I think I addressed previous comments, and as previously mentioned I did rename 'start-up/down' to just 'up/down', which I think makes more sense. I also updated the commit message to clarify that the default behavior in handling the IFF_UP flag is changed, to only set it once for each (re)configuration, instead of every time link_joined() was called.

poettering commented:
lgtm now. I guess @yuwata might want to be the one doing a final review/merge though.

yuwata commented:
Hmm, all CentOS CIs are unhappy. Also, could you rebase this, as I merged #17392?

ddstreet commented:
hmm, well systemd-networkd-tests.py is failing a few tests now, with systemd-networkd-wait-online timing out. I'll run the tests locally to see what i added that's causing the failures.

ddstreet commented:
last push was just a rebase, i'm still looking at what's causing the failures

ddstreet commented:
ok i think it was failing because i had the per-link activation flag in the link->network, which it should not have been, i moved it back to link->activated so it should correctly be a per-link flag now. Also 'networkctl reconfigure' should still work to re-apply the activation policy.

sil2100 commented:
Hello! I wanted to poke about the status of this PR. CI wise it seems to look good - there is one bionic-i386 CI failure which seems to be a timeout issue, possibly 'working' after a re-run. Are there any concerns regarding the design or implementation? Is there anything that can be done to get it merged? Thank you!

yuwata commented:
@sil2100 v247-rc1 was released, and I guess we can release -rc2 near future. So, I'd like to make final review after v247-final released.

poettering commented on src/libsystemd/sd-network/sd-network.c:
```diff
@@ -212,6 +212,10 @@ _public_ int sd_network_link_get_required_operstate_for_online(int ifindex, char
 }
 
+_public_ int sd_network_link_get_activation_policy(int ifindex, char **policy) {
+        return network_link_get_string(ifindex, "ACTIVATION_POLICY", policy);
```
hmm, if the field is missing I'd synthesize "UP" here, i.e. the only supported policy in the status quo ante.

i mean, this doesn't matter, as we don't really support mixing client and server, but it might still be nice to add, because it's easy

ddstreet replied:
ack, i'll update in the next push, thanks

poettering commented:
Still looks good to me, btw. Still leaving the final merge to @yuwata though

yuwata commented on src/network/networkd-network.c:
```diff
@@ -318,6 +339,7 @@ int network_load_one(Manager *manager, OrderedHashmap **networks, const char *fi
 
                 .required_for_online = true,
                 .required_operstate_for_online = LINK_OPERSTATE_RANGE_DEFAULT,
+                .activation_policy = _ACTIVATION_POLICY_INVALID,
```
I'd like to set `ACTIVATION_POLICY_UP` here. If we do not need to detect a setting is actually specified or not, then we should set the default value here.

ddstreet replied:
the only issue with setting it directly is in the check for BindCarrier:

```
        if (!strv_isempty(network->bind_carrier)) {                                                                                                                                                                                                                           
                if (!IN_SET(network->activation_policy, _ACTIVATION_POLICY_INVALID, ACTIVATION_POLICY_BOUND))                                                                                                                                                                 
                        log_warning("%s: ActivationPolicy=bound is required with BindCarrier=. "                                                                                                                                                                              
                                    "Setting ActivationPolicy=bound.", network->filename);                                                                                                                                                                                    
                network->activation_policy = ACTIVATION_POLICY_BOUND;                                                                                                                                                                                                         

```

that currently avoids issuing the conflict warning if ActivationPolicy was unset; but if we set it to its default, we can no longer tell if it was unset or if it was set to 'up', so the warning would either need to be suppressed when someone configures both BindCarrier= and ActivationPolicy=up, or the warning would need to be logged (incorrectly) when someone configures only BindCarrier=.

Do you think we should just drop the warning completely?

yuwata replied:
Ah, indeed. Please ignore my comment here.

yuwata commented on src/network/networkd-network.c:
```diff
@@ -241,6 +238,30 @@ int network_verify(Network *network) {
+        } else if (network->activation_policy == ACTIVATION_POLICY_BOUND) {
+                log_warning("%s: ActivationPolicy=bound requires BindCarrier=. "
+                            "Ignoring ActivationPolicy=bound.", network->filename);
+                network->activation_policy = _ACTIVATION_POLICY_INVALID;
```
And then set `ACTIVATION_POLICY_UP`.

ddstreet replied:
ack

yuwata commented on src/network/networkd-network.c:
```diff
@@ -241,6 +238,30 @@ int network_verify(Network *network) {
+                network->activation_policy = _ACTIVATION_POLICY_INVALID;
+        }
+
+        if (network->activation_policy == _ACTIVATION_POLICY_INVALID)
+                network->activation_policy = ACTIVATION_POLICY_UP;
```
So, we can drop this.

yuwata commented on src/network/networkd-network.c:
```diff
@@ -1252,3 +1274,15 @@ static const char* const ipv6_link_local_address_gen_mode_table[_IPV6_LINK_LOCAL
 DEFINE_CONFIG_PARSE_ENUM(config_parse_ipv6_link_local_address_gen_mode, ipv6_link_local_address_gen_mode, IPv6LinkLocalAddressGenMode, "Failed to parse IPv6 link local address generation mode");
+
+static const char* const activation_policy_table[_ACTIVATION_POLICY_MAX] = {
+        [ACTIVATION_POLICY_UP] = "up",
+        [ACTIVATION_POLICY_ALWAYS_UP] = "always-up",
+        [ACTIVATION_POLICY_MANUAL] = "manual",
+        [ACTIVATION_POLICY_ALWAYS_DOWN] = "always-down",
+        [ACTIVATION_POLICY_DOWN] = "down",
+        [ACTIVATION_POLICY_BOUND] = "bound",
```
Not a matter, but I'd like to align these entries.

ddstreet replied:
ack

yuwata commented on test/test-network/systemd-networkd-tests.py:
```diff
@@ -2710,6 +2711,49 @@ def test_bind_carrier(self):
         self.wait_operstate('test1', 'routable')
 
+    def test_activation_policy(self):
+        for test in ['up', 'always-up', 'manual', 'always-down', 'down', '']:
```
How about to use `self.subTest()`?

ddstreet replied:
it does use subTest() in the up/down/up/down iterations further down in the test, but let me check if it's possible to nest subTests

ddstreet replied:
the nested subTest() calls seem to work great, thanks for the suggestion!

yuwata commented on src/network/networkctl.c:
```diff
@@ -2066,6 +2066,16 @@ static int link_status_one(
+                r = table_add_many(table,
+                                   TABLE_EMPTY,
+                                   TABLE_STRING, "Activation Policy:",
+                                   TABLE_STRING,  activation_policy);
```
Doubled white space after comma. Is it intended to 'align' the field with the above?

ddstreet replied:
wow, good eye! I don't think I intended it for alignment, I'm pretty sure it was just a typo :) will fix in the next push

yuwata reviewed (commented):
LGTM. Some minor comments. Not necessary to be addressed if you do not agree with.

ddstreet commented:
pushed with rebase and some updates as requested by feedback; I think only remaining feedback I'm working on is:
-whether to set 'up' directly instead of using INVALID value, and what to do about bindcarrer conflict warning
-using (nested) subtest in testcase

ddstreet commented:
I think this addresses all the remaining feedback, if the ci passes PTAL
